README.md
<p align="center">
  <img width="480" height="144" src="https://raw.githubusercontent.com/robertcorponoi/graphics/master/godot-rust-cli-upgrader/logo/godot-rust-cli-upgrader-logo.png">
</p>

<h1 align="center">Godot Rust CLI Upgrader</h1>

<p align="center">A CLI tool to help you upgrade your Godot Rust CLI project between versions that introduce breaking changes.</p>

[![Build Status](https://www.travis-ci.com/robertcorponoi/godot-rust-cli-upgrader.svg?branch=main)](https://www.travis-ci.com/robertcorponoi/godot-rust-cli-upgrader)
![Crates.io](https://img.shields.io/crates/v/godot-rust-cli-upgrader)
![Crates.io](https://img.shields.io/crates/d/godot-rust-cli-upgrader)
![Crates.io](https://img.shields.io/crates/l/godot-rust-cli-upgrader)
[![Discord](https://img.shields.io/discord/853728834519040030.svg?label=&logo=discord&logoColor=ffffff&color=7389D8&labelColor=6A7EC2)](https://discord.gg/kr9EkBp7)

## Documentation

Note: If you have any questions about how to use this CLI tool or want to report a bug, feel free to open an issue in the GitHub or the [Discord](https://discord.gg/kr9EkBp7).

Also keep in mind that the main branch will usually be ahead of the version on [crates.io](https://crates.io/crates/godot-rust-cli-upgrader).

**Table of Contents**

- [Introduction](#introduction)
- [Installation](#installation)
- [Step By Step](#step-by-step)
- [API](#api)
    - [upgrade](#upgrade)
    - [version](#version)
- [Questions](#questions)
- [License](#license)

## Introduction

Since [Godot Rust CLI](https://github.com/robertcorponoi/godot-rust-cli) is going through breaking changes pre-v1.0.0, Godot Rust CLI Upgrader was created to help alleviate some of the pain from upgrading between versions. 

## Installation

To install Godot Rust CLI Upgrader, use:

```sh
cargo install godot-rust-cli-upgrader
```

To upgrade your version of Godot Rust CLI Upgrader to the latest version, use:

```sh
cargo install --force godot-rust-cli-upgrader
```

### Step-by-Step

This guide will walk you through upgrading your Godot Rust CLI library to the latest version. If you would rather just look at the API, check it out [below](#api).

1. The first step is to navigate to the library created by Godot Rust CLI. This should be the directory with the configuration file (either `project.toml`, `godot-rust-cli.toml`, or `godot-rust-cli.json`).

2. Figure out what version you want to upgrade to. This should always be the latest version, which at the time of writing this is v0.3.x, but you have a choice to upgrade to a different version. For this guide, we'll be upgrading to to v0.3.x.

**Note:** Godot Rust CLI Upgrader works with major and minor versions but it doesn't care about patches. The only reason we even use minor versions is because pre-v1.0.0 there's going to be breaking changes. This means that when we upgrade, we specify the patch as x like "0.2.x" or "0.3.x".

3. Now that we've decided we want to upgrade to version 0.3.x, we can use the upgrade command like so:

```sh
godot-rust-cli-upgrader upgrade "0.3.x"
```

4. At this point the CLI should notify you of the steps it is taking until it is done. You'll notice that if you're upgrading through multiple versions, like from 0.1.x to 0.3.x, it'll upgrade the library from 0.1.x to 0.2.x first, and then from 0.2.x to 0.3.x. When it's finished you should be able to use your library with the latest version Godot Rust CLI.

## API

### upgrade

Upgrades the version of the library to be compatible with the specified version of Godot Rust CLI.

**Usage:**

```sh
godot-rust-cli-upgrader upgrade <version>
```

where:

- `version` is the version to upgrade the library to. Currently the available versions are: `"0.2.x"` and `"0.3.x"`.

**Examples:**

Upgrading the library to version `"0.3.x"`:

```sh
godot-rust-cli-upgrader upgrade "0.3.x"
```

### version

Returns the version of Godot Rust CLI Upgrader.

**Usage:**

```sh
godot-rust-cli-upgrader --version
```

## Questions

Check out the [Discord](https://discord.gg/kr9EkBp7) to ask any questions or concerns about the cli or Godot + Rust in general.

## License

[MIT](./LICENSE)


CHANGELOG.md
## 0.1.0 / 2021-06-18
- Initial release

tests/upgrade.rs
use assert_cmd::prelude::*;

use std::env::set_current_dir;
use std::error::Error;
use std::fs::read_to_string;
use std::fs::remove_dir_all;
use std::path::Path;
use std::process::Command;

mod test_utilities;
use test_utilities::{create_toml_config, ensure_correct_dir, ConfigToml, ConfigV03X};

/// Upgrades a v0.1.x config to v0.2.x.
#[test]
fn upgrade_config_v_0_1x_to_v_0_2x() -> Result<(), Box<dyn Error>> {
    ensure_correct_dir();

    create_toml_config(false);

    set_current_dir("platformer_modules")?;

    let config_str = read_to_string("project.toml")?;
    let config_toml: ConfigToml = toml::from_str(&config_str)?;

    // 1. Assert the original configuration is correct.
    assert_eq!(config_toml.godot_project_name, "platformer");
    assert_eq!(config_toml.modules, vec!["Player", "MainScene"]);

    // 2. Assert that the upgrade command was successful.
    let mut cmd_upgrade = Command::new("cargo");
    cmd_upgrade
        .arg("run")
        .arg("--manifest-path=../../Cargo.toml")
        .arg("upgrade")
        .arg("0.2.x");
    cmd_upgrade.assert().success();

    // 3. Assert that the old configuration file doesn't exist anymore.
    let old_config_path = Path::new("project.toml");
    assert_eq!(old_config_path.exists(), false);

    // 4. Assert that the new configuration file exists.
    let new_config_path = Path::new("godot-rust-cli.toml");
    assert_eq!(new_config_path.exists(), true);

    let new_config_str = read_to_string(new_config_path)?;
    let new_config_toml: ConfigToml = toml::from_str(&new_config_str)?;

    // 5. Assert that the new configuration is correct.
    assert_eq!(new_config_toml.godot_project_name, "platformer");
    assert_eq!(new_config_toml.modules, vec!["Player", "MainScene"]);

    set_current_dir("../")?;

    remove_dir_all("platformer_modules")?;

    Ok(())
}

/// Upgrades a v0.1.x config to v0.3.x.
#[test]
fn upgrade_config_v_0_1x_to_v_0_3x() -> Result<(), Box<dyn Error>> {
    ensure_correct_dir();

    create_toml_config(false);

    set_current_dir("platformer_modules")?;

    let config_str = read_to_string("project.toml")?;
    let config_toml: ConfigToml = toml::from_str(&config_str)?;

    // 5. Assert that the configuration is correct.
    assert_eq!(config_toml.godot_project_name, "platformer");
    assert_eq!(config_toml.modules, vec!["Player", "MainScene"]);

    // 2. Assert that the upgrade command was successful.
    let mut cmd_upgrade = Command::new("cargo");
    cmd_upgrade
        .arg("run")
        .arg("--manifest-path=../../Cargo.toml")
        .arg("upgrade")
        .arg("0.3.x");
    cmd_upgrade.assert().success();

    // 3. Assert that the old configuration file doesn't exist anymore.
    let old_config_path = Path::new("project.toml");
    assert_eq!(old_config_path.exists(), false);

    // 4. Assert that the new configuration file exists.
    let new_config_path = Path::new("godot-rust-cli.json");
    assert_eq!(new_config_path.exists(), true);

    let new_config_str = read_to_string(new_config_path)?;
    let new_config_json: ConfigV03X = serde_json::from_str(&new_config_str)?;

    // 1. Assert the original configuration is correct.
    assert_eq!(
        new_config_json.name,
        serde_json::json!("platformer_modules")
    );
    assert_eq!(
        new_config_json.godot_project_name,
        serde_json::json!("platformer")
    );
    assert_eq!(new_config_json.is_plugin, serde_json::json!(false));
    assert_eq!(new_config_json.modules[0], "Player");
    assert_eq!(new_config_json.modules[1], "MainScene");

    set_current_dir("../")?;

    remove_dir_all("platformer_modules")?;

    Ok(())
}

/// Upgrades a v0.2.x config to v0.3.x.
#[test]
fn upgrade_config_v_0_2x_to_v_0_3x() -> Result<(), Box<dyn Error>> {
    ensure_correct_dir();

    create_toml_config(true);

    set_current_dir("platformer_modules")?;

    let config_str = read_to_string("godot-rust-cli.toml")?;
    let config_toml: ConfigToml = toml::from_str(&config_str)?;

    // 5. Assert that the configuration is correct.
    assert_eq!(config_toml.godot_project_name, "platformer");
    assert_eq!(config_toml.modules, vec!["Player", "MainScene"]);

    // 2. Assert that the upgrade command was successful.
    let mut cmd_upgrade = Command::new("cargo");
    cmd_upgrade
        .arg("run")
        .arg("--manifest-path=../../Cargo.toml")
        .arg("upgrade")
        .arg("0.3.x");
    cmd_upgrade.assert().success();

    // 3. Assert that the old configuration file doesn't exist anymore.
    let old_config_path = Path::new("project.toml");
    assert_eq!(old_config_path.exists(), false);

    // 4. Assert that the new configuration file exists.
    let new_config_path = Path::new("godot-rust-cli.json");
    assert_eq!(new_config_path.exists(), true);

    let new_config_str = read_to_string(new_config_path)?;
    let new_config_json: ConfigV03X = serde_json::from_str(&new_config_str)?;

    // 1. Assert the original configuration is correct.
    assert_eq!(
        new_config_json.name,
        serde_json::json!("platformer_modules")
    );
    assert_eq!(
        new_config_json.godot_project_name,
        serde_json::json!("platformer")
    );
    assert_eq!(new_config_json.is_plugin, serde_json::json!(false));
    assert_eq!(new_config_json.modules[0], "Player");
    assert_eq!(new_config_json.modules[1], "MainScene");

    set_current_dir("../")?;

    remove_dir_all("platformer_modules")?;

    Ok(())
}


tests/test_utilities.rs
use serde::{Deserialize, Serialize};
use std::fs::{create_dir, write};
use std::path::Path;

/// The structure of the v0.1.x and v0.2.x configuration file.
#[derive(Serialize, Deserialize)]
pub struct ConfigToml {
    pub godot_project_name: String,
    pub modules: Vec<String>,
}

/// The structure of the v0.3.x configuration file.
#[derive(Serialize, Deserialize)]
pub struct ConfigV03X {
    pub name: String,
    pub godot_project_name: String,
    pub is_plugin: bool,
    pub modules: Vec<String>,
}

/// Some tests need to change directory to run correctly so this function is
/// called after those tests to go back to the tests directory that every
/// function expects to be in.
#[allow(dead_code)]
pub fn ensure_correct_dir() {
    let current_dir = std::env::current_dir().unwrap();
    let current_dir_basename = current_dir.file_stem().unwrap();

    if current_dir_basename != "tests" {
        std::env::set_current_dir("tests").expect("Unable to change to tests directory");
    }
}

/// Creates a directory with a v0.1.x configuration file.
///
/// # Arguments
///
/// `is_v_0_2x` - Indicates whether the configuration file is v0.2.x or not.
#[allow(dead_code)]
pub fn create_toml_config(is_v_0_2x: bool) {
    let config = ConfigToml {
        godot_project_name: "platformer".to_owned(),
        modules: vec!["Player".to_owned(), "MainScene".to_owned()],
    };
    let config_toml = toml::to_string_pretty(&config).unwrap();

    let config_dir_path = Path::new("platformer_modules");
    let config_path = if is_v_0_2x {
        config_dir_path.join("godot-rust-cli.toml")
    } else {
        config_dir_path.join("project.toml")
    };

    create_dir(config_dir_path).unwrap();
    write(config_path, config_toml).unwrap();
}

/// Creates a directory with a v0.3.x configuration file.
#[allow(dead_code)]
pub fn create_v_0_3x_config() {
    let config = ConfigV03X {
        name: "platformer_modules".to_owned(),
        godot_project_name: "platformer".to_owned(),
        is_plugin: false,
        modules: vec!["Player".to_owned(), "MainScene".to_owned()],
    };
    let config_json = serde_json::to_string_pretty(&config).unwrap();

    let config_dir_path = Path::new("platformer_modules");
    let config_path = config_dir_path.join("godot-rust-cli.json");

    create_dir(config_dir_path).unwrap();
    write(config_path, config_json).unwrap();
}


src/main.rs
/// Godot Rust CLI Upgrader - A tool used to upgrade your Godot Rust CLI project
/// to be compatible with the latest version of Godot Rust CLI.
///
/// Since Godot Rust CLI undergoes breaking changes pre-1.0.0 and will naturally
/// undergo breaking changes between major versions, this tool helps you upgrade
/// your project so that you don't have to start over.
use structopt::StructOpt;

mod upgrade;

#[path = "./utils/config.rs"]
mod config_utils;
#[path = "./utils/log.rs"]
mod log_utils;

/// Describes the commands available to the user.
#[derive(Debug, StructOpt)]
#[structopt(
    about = "Provides an easy way to upgrade your Godot Rust CLI project to be compatible with the latest version of Godot Rust CLI"
)]
enum GodotRustCliUpgrader {
    /// Upgrades the provided library to be compatible with the latest version
    /// of Godot Rust CLI. This command has to be run from within the library
    /// to upgrade.
    ///
    /// Available versions to upgrade to: "0.2.x", "0.3.x".
    ///
    /// Usage: godot-rust-cli-upgrader --version 0.3.x
    Upgrade {
        /// The version to upgrade the library to.
        #[structopt()]
        version: String,
    },
}

fn main() {
    match GodotRustCliUpgrader::from_args() {
        GodotRustCliUpgrader::Upgrade { version } => upgrade::upgrade(&version),
    }
}


src/upgrade.rs
use lazy_static::lazy_static;
use std::collections::HashMap;

use crate::config_utils::{
    get_config_version, get_version_as_enum_variant, upgrade_config_v0_1_to_v0_2,
    upgrade_config_v0_2_to_v0_3, Version,
};
use crate::log_utils::{log_error_to_console, log_info_to_console, log_success_to_console};

lazy_static! {
    /// A HashMap that maps versions to the functions that are used to upgrade the
    /// configuration from the previous version to that version.
    pub static ref VERSION_UPGRADE_HANDLER: HashMap<Version, fn() -> ()> = {
        let mut version_upgrade_handler = HashMap::new();
        version_upgrade_handler.insert(Version::V02X, upgrade_config_v0_1_to_v0_2 as fn() -> ());
        version_upgrade_handler.insert(Version::V03X, upgrade_config_v0_2_to_v0_3 as fn() -> ());

        version_upgrade_handler
    };
}

/// Upgrades the library in the current directory to be compatible with the
/// specified version of Godot Rust CLI.
///
/// # Arguments
///
/// `version` - The version to upgrade the library to.
pub fn upgrade(version: &str) {
    let version_as_enum_value = get_version_as_enum_variant(version);
    if !is_version_in_accepted_versions(&version_as_enum_value) {
        log_error_to_console("The provided version cannot be upgraded to. Check the documentation for valid versions that can be upgraded to.");
    }

    log_info_to_console("Retrieving current version...");
    let current_config_version = get_config_version();

    // Get the index of the current version of the configuration and the
    // desired version to upgrade to in the versions vec.
    let versions = get_versions();
    let index_of_current_config_version_in_versions = &versions
        .iter()
        .position(|i| i == &current_config_version)
        .expect("Unable to get index of current configuration version while upgrading");
    let index_of_version_to_upgrade_to_in_versions = &versions
        .iter()
        .position(|i| i == &version_as_enum_value)
        .expect("Unable to get index of the version to upgrade to while upgrading");

    for (pos, v) in versions.iter().enumerate() {
        if pos < *index_of_current_config_version_in_versions || v == &current_config_version {
            // If the version is at or below the index of our current version
            // then we can continue past it.
            continue;
        }

        // Run the function associated with the version to upgrade the
        // configuration to.
        let upgrade_fn_to_run = VERSION_UPGRADE_HANDLER
            .get(&v)
            .expect("Unable to run the upgrade handler for the specified version.");
        upgrade_fn_to_run();

        if pos + 1 > *index_of_version_to_upgrade_to_in_versions {
            // If the version is past the upgrade index then we can return as
            // we are done upgrading.
            break;
        }
    }

    log_success_to_console("Upgrade complete");
}

/// Indicates whether the provided version to upgrade to is on the list of
/// supported versions.
///
/// # Arguments
///
/// `version` - The version to upgrade the library to.
fn is_version_in_accepted_versions(version: &Version) -> bool {
    let accepted_versions = get_versions();
    if accepted_versions.iter().any(|i| &i == &version) {
        return true;
    }
    return false;
}

/// Returns the versions used while upgrading the configuration.
fn get_versions() -> Vec<Version> {
    return vec![Version::V01X, Version::V02X, Version::V03X];
}


src/utils/config.rs
use crate::log_utils::{log_info_to_console, log_success_to_console};

use serde::{Deserialize, Serialize};
use std::env::current_dir;
use std::fs::{read_to_string, remove_file, rename, write};
use std::path::Path;

/// The v0.1.0-v0.1.3 configuration object.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ConfigV01 {
    pub godot_project_name: String,
    pub modules: Vec<String>,
}

/// Defines the versions that are used throughout the program to define the
/// current version of the library and versions that can be upgraded to.
#[derive(Debug, PartialEq, Eq, Hash)]
pub enum Version {
    V01X,
    V02X,
    V03X,
}

/// Returns the version of the config based on either the cli version in the
/// config or other context pre-cli version.
pub fn get_config_version() -> Version {
    let current_dir_path = current_dir()
        .expect("Unable to get current directory while getting the configuration version.");

    if current_dir_path.join("project.toml").exists() {
        // If the library has a configuration file named `project.toml` then
        // we know that it is a v0.1.x configuration.
        return Version::V01X;
    } else if current_dir_path.join("godot-rust-cli.toml").exists() {
        // If the library has a configuration file named `godot-rust-cli.toml`
        // then we know that it is a v0.2.x configuration.
        return Version::V02X;
    } else {
        // For v0.3.x forward, Godot Rust CLI switched to a JSON config.
        // However, in v0.3.x versions, the cli version property doesn't exist
        // exist in the config so we know if it's v0.3.x depending on that.
        let config = read_to_string("godot-rust-cli.json")
            .expect("Unable to read configuration file while getting the configuration version.");
        let config_json: serde_json::Value = serde_json::from_str(&config).expect(
            "Unable to parse configuration file to JSON while getting the configuration version.",
        );

        let cli_version = &config_json["cliVersion"];
        let cli_version_str = cli_version
            .as_str()
            .expect("Unable to parse cli version while getting the configuration version.");
        if cli_version.is_string() {
            return get_version_as_enum_variant(&cli_version_str);
        } else {
            return Version::V03X;
        }
    }
}

/// Returns the version as a Version enum variant.
///
/// `version_to_convert` - The version to convert to an enum variant.
pub fn get_version_as_enum_variant(version_to_convert: &str) -> Version {
    return match version_to_convert {
        "0.2.x" => Version::V02X,
        "0.3.x" => Version::V03X,
        _ => Version::V03X,
    };
}

/// Updates the configuration from v0.1.x to v0.2.x by renaming the
/// configuration file from `project.toml` to `godot-rust-cli.toml`.
pub fn upgrade_config_v0_1_to_v0_2() {
    log_info_to_console("Starting configuration upgrade from v0.1.x to v0.2.x...");

    let original_config_path = Path::new("project.toml");
    let new_config_path = Path::new("godot-rust-cli.toml");
    rename(original_config_path, new_config_path)
        .expect("Unable to rename configuration file while upgrading from v0.1.x to v0.2.x");

    log_success_to_console("Finished upgrading configuration from v0.1.x to v0.2.x");
}

/// Updates the configuration from v0.2.x to 0.3.x by replacing the
/// toml contents and file type with json.
pub fn upgrade_config_v0_2_to_v0_3() {
    log_info_to_console("Starting configuration upgrade from v0.2.x to v0.3.x...");

    let original_config_path = Path::new("godot-rust-cli.toml");
    let new_config_path = Path::new("godot-rust-cli.json");

    let toml_config_as_str = read_to_string(&original_config_path)
        .expect("Unable to read original configuration while upgrading from v0.2.x to v0.3.x.");
    let toml_config: toml::Value = toml::from_str(&toml_config_as_str)
        .expect("Unable to parse original configuration while upgrading from v0.2.x to v0.3.x.");
    let godot_project_name = toml_config["godot_project_name"]
        .as_str()
        .expect("Unable to convert godot project name while reading configuration.");

    let current_dir = std::env::current_dir()
        .expect("Unable to get current directory while upgrading from v0.2.x to v0.3.x.");
    let library_name = current_dir
        .file_stem()
        .expect("Unable to get library name while upgrading from v0.2.x to v0.3.x")
        .to_str()
        .expect("Unable to get library name while upgrading from v0.2.x to v0.3.x");

    let json_config = serde_json::json!({
        "name": library_name,
        "godot_project_name": godot_project_name,
        "is_plugin": false,
        "modules": toml_config["modules"].as_array().expect("Unable to convert modules while reading configuration.")
    });
    let json_config_as_string = serde_json::to_string_pretty(&json_config)
        .expect("Unable to convert new configuration to JSON string");

    write(new_config_path, json_config_as_string)
        .expect("Unable to create new configuration file while upgrading from v0.2.x to v0.3.x.");

    remove_file(original_config_path)
        .expect("Unable to delete old configuration file while upgrading from v0.2.x to v0.3.x.");

    log_success_to_console("Finished upgrading configuration from v0.2.x to v0.3.x");
}
